feat: v5 release blog post (#1616)
expressjs/expressjs.com: _posts/2024-09-12-v5-release.md

While we weren't able to simultaneously release this blog post, our changelog, and the documentation, we felt it was most important to have a secure and stable release.

As soon as possible, we'll provide more details on our long-term support (LTS) plans, including when the release will move from `next` to `latest`. For now, if you are uncomfortable being on the bleeding edge (even if it is a rather dull edge) then you should wait to upgrade until the release is on tagged as `latest`. That said, we look forward to working with you to address any bugs you encounter as you upgrade.

## Breaking changes

> The v5 release has the minimum possible number of breaking changes, listed here in order of impact to applications. However, there are also number of subtle changes: for details, see the changelog.

Reviewer: I think it's better to recommend checking the migration guide instead of the changelog

Reviewer 2: Good catch, I made that change.  I also changed the other link to the migration guide to be canonical with lang variable.

---
title: "Introducing Express v5: A New Era for Node.js Framework"
tags: releases
author: Wes Todd and the Express Technical Committee
description: Announcing the release of Express version 5
---

> Ten years ago the [Express v5 release pull request](https://github.com/expressjs/express/pull/2237) was opened, and now at long last it's been merged and published!  We want to recognize the work of all our contributors, especially [Doug Wilson](https://github.com/dougwilson), who spent the last ten years ensuring Express was the most stable project around. Without his contributions and those of many others, this release could not have happened.

Reviewer 3: ```suggestion
Ten years ago (July 2014) the [Express v5 release pull request](https://github.com/expressjs/express/pull/2237) was opened, and now at long last it's been merged and published!  

We want to recognize the work of all our contributors, especially [Doug Wilson](https://github.com/dougwilson), who spent the last ten years ensuring Express was the most stable project around. Without his contributions and those of many others, this release could not have happened.
```

> Eight months ago we went public with a plan to move [Express forward](https://github.com/expressjs/discussions/issues/160). This plan included re-committing to the governance outlined years ago and adding more contributors to help kickstart progress.  Many people may not realize that robust project governance is critical to the health of a large open-source project. We want to thank the [OpenJS Foundation Cross Project

Reviewer 3: ```suggestion
Eight months ago we went public with a plan to move [Express forward](https://github.com/expressjs/discussions/issues/160). This plan included re-committing to the governance outlined years ago and adding more contributors to help kickstart progress. Many people may not realize that robust project governance is critical to the health of a large open-source project. We want to thank the [OpenJS Foundation Cross Project
```

Council](https://github.com/openjs-foundation/cross-project-council/) and its members for helping us put together this plan. 

## So what about v5?

> This release is designed to be boring!  That may sound odd, but we've intentionally kept it simple to unblock the ecosystem and enable more impactful changes in future releases. The focus of this release is on dropping old Node.js version support, addressing security concerns, and simplifying maintenance.

Reviewer 3: ```suggestion
This release is designed to be boring! 
That may sound odd, but we've intentionally kept it simple to unblock the ecosystem and enable more impactful changes in future releases. This is also about signaling to the Node.js ecosystem that Express is moving again.
The focus of this release is on dropping old Node.js version support, addressing security concerns, and simplifying maintenance.
```

> Before going into the changes in this release, let's address why it was released v5 on the `next` dist-tag. As part of reviving the project, we started a [Security working group](https://github.com/expressjs/security-wg) and [security triage team](https://github.com/expressjs/security-wg?tab=readme-ov-file#security-triage-team) to address the growing needs around open source supply chain security. We undertook a security audit (more details to come on that) and uncovered some problems that needed to be addressed. Thus, in addition to the "normal" work done in public issues, we also did a lot of security work in private forks. This security work required orchestration when releasing, to ensure the code and CVE reports went out together. You can find a summary of the most recent vulnerabilities patched in [our security release notes](https://expressjs.com/2024/09/29/security-releases.html). 

Reviewer 3: ```suggestion
Before going into the changes in this release, let's address why it was released v5 on the `next` dist-tag. As part of reviving the project, we started a [Security working group](https://github.com/expressjs/security-wg) and [security triage team](https://github.com/expressjs/security-wg?tab=readme-ov-file#security-triage-team) to address the growing needs around open source supply chain security. We undertook a security audit (more details to come on that) and uncovered some problems that needed to be addressed. Thus, in addition to the "normal" work done in public issues, we also did a lot of security work in private forks. 
This security work required orchestration when releasing, to ensure the code and CVE reports went out together. You can find a summary of the most recent vulnerabilities patched in [our security release notes](https://expressjs.com/2024/09/29/security-releases.html). 
```

While we weren't able to simultaneously release v5, this blog post, the changelog, and documentation, we felt it was most important to have a secure and stable release.

As soon as possible, we'll provide more details on our long-term support (LTS) plans, including when the release will move from `next` to `latest`. For now, if you are uncomfortable being on the bleeding edge (even if it is a rather dull edge) then you should wait to upgrade until the release is tagged `latest`. That said, we look forward to working with you to address any bugs you encounter as you upgrade.

## Breaking changes

The v5 release has the minimum possible number of breaking changes, listed here in order of impact to applications. 

- [Ending support for old Node.js versions](#goodbye-nodejs-010-hello-node-18)
- [Changes to path matching and regular expressions](#changes-to-path-matching-and-regular-expressions)
- [Promise support](#promise-support)
- [Body parser changes](#body-parser-changes)
- [Removing deprecated method signatures](#removing-deprecated-method-signatures)

> There are also number of subtle changes: for details, see [Migrating to Express 5]({{site.url}}/{{page.lang}}/guide/migrating-5).

Reviewer 3: ```suggestion
There are also a number of subtle changes: for details, see [Migrating to Express 5]({{site.url}}/{{page.lang}}/guide/migrating-5).
```

We recognize that this might cause difficulty for some enterprises with older or "parked" applications, and because of this we are working on a [partnership with HeroDevs](https://expressjs.com/2024/10/01/HeroDevs-partnership-announcement.html) to offer "never-ending support" that will include critical security patches even after v4 enters end-of-life (more on these plans soon). That said, we strongly suggest that you update to modern Node.js versions as soon as possible.

### Changes to path matching and regular expressions

The v5 releases updates to `path-to-regexp@8.x` from `path-to-regexp@0.x`, which incorporates many years of changes. If you were using any of the 5.0.0-beta releases, a last-minute update which greatly changed the path semantics to [remove the possibility of any ReDoS attacks](https://blakeembrey.com/posts/2024-09-web-redos/). For more detailed changes, [see the `path-to-regexp` readme](https://github.com/pillarjs/path-to-regexp?tab=readme-ov-file#express--4x).

#### No more regex

> This releases no longer supports "sub-expression" regular expressions, for example `/:foo(\\d+)`.

Reviewer 3: ```suggestion
This release no longer supports "sub-expression" regular expressions, for example `/:foo(\\d+)`.
```

This is a commonly-used pattern, but we removed it for security reasons. Unfortunately, it's easy to write a regular expression that has exponential time behavior when parsing input: The dreaded regular expression denial of service (ReDoS) attack. It's very difficult to prevent this, but as a library that converts strings to regular expressions, we are on the hook for such security aspects.

> *How to migrate:* The best approach to prevent ReDoS attacks is to use a robust input validation library. [There are many on `npm`](https://www.npmjs.com/search?q=validate%20express) depending on your needs.  TC member Wes Todd maintains [a middleware-based "code first" OpenAPI library](https://www.npmjs.com/package/@wesleytodd/openapi) for this kind of thing. 

Reviewer 3: ```suggestion
*How to migrate:* The best approach to prevent ReDoS attacks is to use a robust input validation library. [There are many on `npm`](https://www.npmjs.com/search?q=validate%20express) depending on your needs. TC member Wes Todd maintains [a middleware-based "code first" OpenAPI library](https://www.npmjs.com/package/@wesleytodd/openapi) for this kind of thing. 
```

- `req.acceptsEncoding`: Use `req.acceptsEncodings` (plural).
- `req.acceptsLanguage`: Use `req.acceptsLanguages` (plural).
- `res.sendfile` method: Use `res.sendFile` instead.

As a framework, we aim to ensure that the API is as consistent as possible. We've removed these deprecated signatures to make the API more predictable and easier to use. By streamlining each method to use a single, consistent signature, we simplify the developer experience and reduce confusion.

## Migration and security guidance

For developers looking to migrate from v4 to v5, there's a [detailed migration guide]({{site.url}}/{{page.lang}}/guide/migrating-5) to help you navigate through the changes and ensure a smooth upgrade process.

Additionally, we’ve been working hard on a comprehensive [Threat Model](https://github.com/expressjs/security-wg/blob/main/docs/ThreatModel.md) that helps illustrate our philosophy of a "Fast, unopinionated, minimalist web framework for Node.js." It provides critical insights into areas like user input validation and security practices that are essential for safe and secure usage of Express in your applications.

## Our work is just starting

> We see the v5 release as a milestone toward an Express ecosystem that's a stable and reliable tool for companies, governments, educators, and hobby projects.  It is our commitment as the new stewards of the Express project to move the ecosystem forward with this goal in mind. If you want to support this work, which we do on a volunteer basis, please consider supporting the project and its maintainers via [our sponsorship opportunities](https://opencollective.com/express).  

Reviewer 3: ```suggestion
We see the v5 release as a milestone toward an Express ecosystem that's a stable and reliable tool for companies, governments, educators, and hobby projects. It is our commitment as the new stewards of the Express project to move the ecosystem forward with this goal in mind. If you want to support this work, which we do on a volunteer basis, please consider supporting the project and its maintainers via [our sponsorship opportunities](https://opencollective.com/express).  
```

> We have an [extensive working backlog](https://github.com/expressjs/discussions/issues/266) of tasks, PRs, and issues for Express and dependencies.  Naturally, we expect developers will continue to report issues to add to this backlog and open PRs moving forward, and we'll continue to collaborate with the community to triage and resolve them.  We look forward to continuing to improve Express and making it useful for its users across the world.

Reviewer 3: ```suggestion
We have an [extensive working backlog](https://github.com/expressjs/discussions/issues/266) of tasks, PRs, and issues for Express and dependencies. Naturally, we expect developers will continue to report issues to add to this backlog and open PRs moving forward, and we'll continue to collaborate with the community to triage and resolve them.  We look forward to continuing to improve Express and making it useful for its users across the world.
```

This is one of the most common patterns in wide use which we have removed. The unfortunate nature of regular expressions is how easy it is to write one with exponential time behaviors
when parsing input. The dreaded ReDoS. It turns out it is very difficult to prevent users from doing this, and as a library which converts strings to regular expressions, we are on
the hook for the security aspects of this.

*How to migrate:* We recommend using more robust input validation libraries. [There are many on `npm`](https://www.npmjs.com/search?q=validate%20express) depending on your needs.
Shameless plug from the author, I maintain [a middleware based "code first" OpenAPI library](https://www.npmjs.com/package/@wesleytodd/openapi) for this kind of thing. 

#### 2. Splats, optional, and captures oh my. Simplified patterns for common route patterns.

With the removal of regular expression semantics comes other small but impactful changes to how you write your routes.

1. `:name?` becomes `{:name}`. Usage of `{}` for optional parts of your route means you can now do things like `/base{/:optional}/:required` and what parts are actually optional is
   much more explicit.

> 2. `:name*` becomes `*name`. (@blakeembrey to provide more details)

Reviewer 4: ```suggestion
2. `*` becomes `*name`.
```

### Body Parser

We made an assortment of `body-parser` changes:

- Add option to customize the urlencoded body depth with a default value of 32 (@TODO see CVE)
- Remove deprecated `bodyParser()` combination middleware
- `req.body` is no longer always initialized to `{}`
- `urlencoded` parser now defaults `extended` to false
- Added brotli support

### Deprecated API Method Signatures Removed in Express v5

We have removed a number of deprecated API method signatures, many of which were carried over from v3 and deprecated under v4. Below are the changes you need to make:

- `res.redirect('back')` and `res.location('back')`: The magic string `'back'` is no longer supported. Use `req.get('Referrer') || '/'` explicitly instead.
- `res.send(status, body)` and `res.send(body, status)` signatures: Use `res.status(status).send(body)`.
- `res.send(status)` signature: Use `res.sendStatus(status)` for simple status responses, or `res.status(status).send()` for sending a status code with an optional body.
- `res.redirect(url, status)` signature: Use `res.redirect(status, url)`.
- `res.json(status, obj)` and `res.json(obj, status)` signatures: Use `res.status(status).json(obj)`.

> - `res.jsonp(status, obj)` and `res.jsonp(obj, status)` signatures: Use `res.status(status).jsonp(obj)`.

Reviewer 4: updated this list bc we previously were saying

`res.send(body, status) deprecated, use `res.(status, body)` then later said `res.send(status, body) deprecated use res.status(status).send(body)` 

Had that for json, jsonp, and send.

---
title: Express v5

> tags: site-admin

Author: Obviously this is the wrong tag, I just copied this from the first post. What tags do we want to use?

Reviewer 5: `releases`? maybe is a good option. Just copying Node

![Screenshot from 2024-09-13 20-40-43](https://github.com/user-attachments/assets/39913816-fd82-48c6-837d-da02d65067df)

Reviewer 2: I'll make it releases.  Currently, the tags are not even used, so it doesn't actually matter...lol.

> author: Wes Todd

Reviewer: The description could be added to improve SEO
```suggestion
description: <description>
author: Wes Todd
```

Reviewer 2: Added a description.

---

> A lot has happened in the last 10 years, but today we are excited to talk about what has happened in the last 8 months. The Express project has had a renaissance, and we are

Reviewer 6: ```suggestion
Ten years ago the [Express v5 release PR](https://github.com/expressjs/express/pull/2237) was opened. It is now merged and published! There's a lot to be excited about, but we want to recognize the work of all our contributors, especially Doug Wilson who spent the last 10 years ensuring Express was the most stable project around. Without these contributions this release could not have happened.
```

Trying to punch it up a little and avoid starting with "a lot has happened". Also avoid 8 months until next paragraph to keep focus on "released" and congrats. Finally, avoid switching between we/I.

Reviewer 6: I also like the idea of an immediate call to action of "Version 5 has been published and you can install it today: `npm install express@5`!"

Or something like this. Then move on to "The initial PR was opened 10 years ago and there's a lot of be excited about, but the most important thing to mention is the diligent work of all the maintainers and contributors. Special call out to Doug Wilson [...]"

Reviewer 7: @blakeembrey when express 5 will be released, simply doing `npm i express` will automatically install the latest version.
For existing projects, specifying version 5 is essential to do

I think it is better if we mention `npm i express@latest`

excited to finally share with you some of the work we have been doing. As you might be aware, the [v5 release which has been in progress since July 14th
2014](https://github.com/expressjs/express/pull/2237) is now published and merged. There is a lot to talk about in the release, but I want to take a second to recognize the work of
many contributors over the years, especially @dougwilson who tirelessly maintained one of the most stable projects in the ecosystem over the past 10 years. Without the work from so
many, this release would not have happened so if you are among these contributors please give yourself a pat on the back. 

> So what happened 8 months ago? We went public with a proposed plan to move [Express Forward](https://github.com/expressjs/discussions/issues/160). This plan included re-committing

Reviewer 6: ```suggestion
Eight months ago we went public with a plan to move [Express forward](https://github.com/expressjs/discussions/issues/160). This plan included re-committing
```

to the governance we had outlined many years before and added some things to help onboard more contributors to help kickstart progress again. It might not seem important to folks

> who are less involved in Open Source, but project governance is critical to larger projects health. I want to thank the [OpenJS Foundation Cross Project

Reviewer 6: ```suggestion
who are less involved in Open Source, but project governance is critical to larger projects health. We want to thank the [OpenJS Foundation Cross Project
```

Is this OK to third person?

> Council](https://github.com/openjs-foundation/cross-project-council/) and it's members for helping us put together this plan. 

Reviewer 6: ```suggestion
Council](https://github.com/openjs-foundation/cross-project-council/) and its members for helping us put together this plan. 
```

## So what about v5?

> Lets start by mentioning how **boring** this release is meant to be. I know this might seem like an odd thing to say, but it was truly our goal to keep this release as simple as we

Reviewer 6: Worried this whole blog post is reading like a speech. Could just read "This release is designed to be **boring**". Using I vs we again.

> This release is designed to be boring. It's a small jumpstart. It has been kept simple to unblock the ecosystem and allow more impactful changes to come in v6 and beyond. Our focus was on dropping old Node.js version support, addressing security concerns, and simplifying maintenance.

Reviewer 2: Incorporated and reworded somewhat.

could while unblocking ourselves to make larger and more impactful changes in future releases. This means we focused on things like dropping support for older Node.js versions,
addressing long standing security concerns, and updating the projects process to make things more maintainable for maintainers. To most folks these sound pretty **boring**, but to
us this means we can more easily ship future feature releases with the more exciting changes.

> Before I move onto the changes for Express users we need to address the timeline and reason we released v5 when we did and on the `next` dist-tag. As part of reviving the project,

Reviewer 6: Very speech like again, we can just kill the first sentence and clarify things quicker and make it easier to skim. Something like:

> Version 5 has been release as the `next` dist tag. [...]

Two sentences in a row start with "As part".

Reviewer 2: Incorporated, reworded, edited, and simplifed.

uncovered some other problems which needed to be addressed. This meant that in addition to the normal work we would do in public issues, we had many stacked features being worked
on in private forks. These required orchestration when releasing to enure the code and CVE reports went out together. So while we recognize that not having this blog post, our
changelog, and the documentation updated in advance was not ideal, with our limited contributor base we felt it was more important to focus on getting the secure and stable code
released.

We will be publishing more details on our LTS plans, including dates when we plan to move from `next` to `latest`, as soon as we can. But for now, if you are uncomfortable being on
the bleeding edge (even if it is a rather dull edge) then please consider waiting until we move to `latest` to upgrade. That said, we look forward to working with you all as you
upgrade to support and address any bugs you encounter.

## The Breaking Changes

As I said above, we did the minimum number of breaking changes we could. I try here to list them in order of impact to application owners, but there are a fair number of subtle

> chnages burried in here which you should read the changelog for more details on.

Reviewer 6: ```suggestion
changes burried in here which you should read the changelog for more details on.
```

> ### Goodbye Node.js 0.10, Hello Node 18 and up.

Reviewer 6: ```suggestion
### Goodbye Node.js 0.10, Hello Node 18 and Up
```

Consistent title casing. Personally, I'd just drop title case and use sentence case.

We dropped support for Node.js versions lower than 18. This was a long time coming, but it is probably the most important change for us as maintainers of these libraries. Keeping old
Node.js version support was holding us back from many critical performance and maintainability changes. Now that we have dropped these versions we have more stable and maintainable
CI, we can start adopting some newer language and runtime features, and we can drop many dependencies which are no longer required.

> We recognize that this might mean some enterprises have difficulty with older or "parked" applications, and because of this we are working on a partnership with HeroDevs to offer

Reviewer 8: ```suggestion
We recognize that this might mean some enterprises have difficulty with older or "parked" applications, and because of this we are working on a partnership with [HeroDevs](https://www.herodevs.com/) to offer
```

Reviewer: Could it be better to redirect directly to the page that talks about Express? https://www.herodevs.com/support/express-nes

Reviewer 4: ```suggestion
We recognize that this might mean some enterprises have difficulty with older or "parked" applications, and because of this we are working on a [partnership with HeroDevs](https://expressjs.com/2024/10/01/HeroDevs-partnership-announcement.html) to offer
```

Our own blog post is up, so we can link to that directly.

We updated from `path-to-regexp@0.x` to `path-to-regexp@8.x` in v5, needless to say there were many years of changes in this. If you were using any of the 5.0.0-beta releases, we
landed a last minute update which greatly changed the path semantics so we could [remove any ReDoS chances going forward](https://blakeembrey.com/posts/2024-09-web-redos/). For
more detailed changes, [see the `path-to-regexp` readme](https://github.com/pillarjs/path-to-regexp?tab=readme-ov-file#express--4x).

#### 1. No more regex. We dropped support for sub-expression regular expressions (ex. `/:foo(\\d+)`).

This is one of the most common patterns in wide use which we have removed. The unfortunate nature of regular expressions is how easy it is to write one with exponential time behaviors
when parsing input. The dreaded ReDoS. It turns out it is very difficult to prevent users from doing this, and as a library which converts strings to regular expressions, we are on
the hook for the security aspects of this.

*How to migrate:* We recommend using more robust input validation libraries. [There are many on `npm`](https://www.npmjs.com/search?q=validate%20express) depending on your needs.

> Shameless plug from the author, I maintain [a middleware based "code first" OpenAPI library](https://www.npmjs.com/package/@wesleytodd/openapi) for this kind of thing. 

Reviewer 6: Let's not plug this if it's a focus on V5 and our contributions, I think this could be a whole other blog post where you focus on validation inputs. Also another first person switch.

Author: I do not mind first person switchs, it was written by a person lol. That said, I want to give users a clear recommendation, not a wishy wasy "there are options", but since I have never used the other options I cannot vouch for them to make a recommendation. Happy to take a poll of the team and see which ones folks have had success with and recommend a list of them, but I would rather not tell an audience with mixed technical backgrounds "use a validation lib" with no guidance on what that means.

Reviewer 2: I left this in there for now.  I don't think it hurts to provide a cross-reference for those who might need it.

#### 2. Splats, optional, and captures oh my. Simplified patterns for common route patterns.

With the removal of regular expression semantics comes other small but impactful changes to how you write your routes.

1. `:name?` becomes `{:name}`. Usage of `{}` for optional parts of your route means you can now do things like `/base{/:optional}/:required` and what parts are actually optional is
   much more explicit.

> 2. `:name*` becomes `*name`. (@blakeembrey to provide more details)

Reviewer 6: `*` becomes `*name`.

> 3. `:name+` is equivalent to a `*name` and so has been removed

Reviewer 6: Express 4 never supported this.

Author: https://github.com/pillarjs/router/commit/30b42cbee88e7f9a19a75d2bf1538ea14a537963#diff-859f2b16134f796c6f0bb540c54d824024ff05b9722708e13ee78b29476e3471L867

These tests say otherwise?

Author: Oh! Are you saying this was an artifact of the intermediary PRs?

Reviewer 6: The betas, yeah, `path-to-regexp` has the changes in the README here: https://github.com/pillarjs/path-to-regexp?tab=readme-ov-file#express--4x

Reviewer 5: I remove the reference in 9daa4c10dac87b0c25facffd40fea6868120657a. Please revert the commit if that is not correct.

In Express v4, using regex capture groups you could get numerical parameters (ex. `/user(s?)` => `req.params[0] === 's'`). Now all params must be named. Along with requiring a
name, we now support all valid JavaScript identifiers or quoted (ex. `/:"this"`).

### Promise Support

This one may be a bit contentious, but I `Promise` we are moving in the right direction. We have added support for returned *rejected* promises from middleware raising errors. This
*does not include* calling `next` from returned *resolved* promises. There are a lot of edge cases in old express apps around expectations of `Promise` behavior, and before we can
run we need to walk. For most folks, this means you can now write middleware like the following:

```javascript
app.use(async (req, res, next) => {
  req.locals.user = await getUser(req);
  next();
});
```

Notice that we use `async/await` and the `getUser` call may throw (user does not exist, db is down, etc), but we still call `next` if it is successful. We dont need to catch the
error in line anymore if we want to rely on error handling middleware instead because the router will now catch the rejected promise and treat that as calling `next(err)`.

> *editorial note:* Error handling is a huge topic, but one hill I will die on is that errors should be handled as close to the error site as possible. So while this is now handled

Reviewer 6: Remove this, focus on V5 changes.

Author: I am pretty against examples like the one above without a note that this is a bad practice. But I also want to make sure we dont make an overly complicated example either. Would love to hear others opinions on this kind of thing.

Reviewer 2: Reworded it a bit. IMO it's OK to note best practices, even if unrelated and in a blog post.

in the router, I would strongly urge you to catch this kind of error within the middleware and handle it without relying on separate error handling middleware.

### Body Parser

We made an assortment of `body-parser` changes:

- Add option to customize the urlencoded body depth with a default value of 32 (@TODO see CVE)
- Remove deprecated `bodyParser()` combination middleware
- `req.body` is no longer always initialized to `{}`
- `urlencoded` parser now defaults `extended` to false
- Added brotli support

> ### Removed APIs

Reviewer 9: Since its a blog and probably the interested ones will read it, would it be a good idea to include a reason or link to something so that users can understand why the removed APIs were removed?

Reviewer 5: I like the idea but I think that will make the blog post longer to read. Also the links probably will be only links to commits so it will be hard to contextualize, maybe we can include a sentence to summarize the strategy that we follow as a framework.

Reviewer 5: I added an explanation in eea125540482980f218c67a8a8cc355c15120cba

> ### Removed APIs

Reviewer 4: ```suggestion
### Removed Deprecated APIs
```

We have removed a bunch of apis which were primarily from v3.

- `res.redirect('back')` and `res.location('back')` are no longer a supported magic string, explicitly use `req.get('Referrer') || '/'`
- `res.send(status, body)` - use `res.status(status).send(body)`
- `res.redirect(url, status)` - use `res.redirect(status, url)`
- `res.jsonp(status, obj)` - use `res.status(status).jsonp(obj)`
- `res.json(status, obj)` - use `res.status(status).json(obj)`
- `app.param(fn)` - use `req.params`, `req.body` or `req.query` instead
- `app.del` - use `app.delete`
- `req.acceptsCharset` - use `req.acceptsCharsets`
- `req.acceptsEncoding` - use `req.acceptsEncodings`
- `req.acceptsLanguage` - use `req.acceptsLanguages`
- `res.json(obj, status)` signature - use `res.json(status, obj)`
- `res.jsonp(obj, status)` signature - use `res.jsonp(status, obj)`

> - `res.send(body, status)` signature - use `res.send(status, body)`

Reviewer 9: I think I have a understanding gap here, please help me to understand. Line 129 says,

```md
`res.send(body, status)` signature - use `res.send(status, body)`
```

But line 119 says,
```md
`res.send(status, body)` - use `res.status(status).send(body)`
```

And since `res.send(status, body)` no longer works, it might be confusing to suggest that? Sames goes for other few methods as I see.

Reviewer 4: This is a good callout. It was pulled from the changelog which evolved as the betas did. These are deprecated signatures. `res.send()` no longer accepts a status input.

- `res.send(status)` signature - use `res.sendStatus(status)`
- `res.sendfile` - use `res.sendFile` instead

> ## Our work is just starting

Reviewer 5: Maybe we can mention the v5.x backlog? https://github.com/expressjs/discussions/issues/266

Author: Oh yeah great idea!

Reviewer 2: Added.
